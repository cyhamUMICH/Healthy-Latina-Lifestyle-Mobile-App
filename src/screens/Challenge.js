import React, { useState, useEffect } from "react";
import { View, Text, Image, ScrollView, CheckBox, ButtonGroup} from "react-native";
import Controller from "../components/Controller";
import PlayerSlider from "../components/PlayerSlider";
import Tags from '../components/Tags';
import { styles } from '../styles/Styles';
import { Button } from 'react-native-elements';
import ContentList from '../components/ContentList';
import SetFeatured from "../components/SetFeatured";
import LoadingSpinner from '../components/LoadingSpinner';
import * as firebase from 'firebase/app';
import "firebase/firestore";
import "firebase/storage";
import ChallengeButtons from "../components/ChallengeDayList";
import { Checkbox } from 'react-native-paper';
import ChallengeDayList from '../components/ChallengeDayList';
import defaultImage from '../../assets/logo-icon.png';


const Challenge = ({route}, props) => {
  
  const item = route.params.item;
  const navigation = route.params.navigation;
  const theContentID = item.contentID;
  
  const [data, setData] = useState([]);
  const [isLoaded, setIsLoaded] = useState(false);
  
  const theStartDate = item.startDate.toDate().toString().slice(4, 15);
  const theEndDate = item.endDate.toDate().toString().slice(4, 15);

  useEffect(() => {

    const fetchList = async () => {
      const dbh = firebase.firestore();
      
      dbh.collection("challengeDays").orderBy('date').get()
      .then((querySnapshot) => {

        if (querySnapshot.size == 0) {
          setIsLoaded(true);
        }
        else {
          let countChallenges = 0;
      
          querySnapshot.forEach((doc) => {
            let newDoc = doc.data();
            newDoc.contentID = doc.id;

            const chalref = newDoc.challenge.id;       
    
            // https://firebase.google.com/docs/storage/web/download-files
            let storage = firebase.storage();
            let pathReference = storage.ref(newDoc.title);

           if(chalref == theContentID){
             
              countChallenges++;
        
              setData(oldList => [...oldList, newDoc]);

            }

            if(countChallenges == item.numDays){

              setIsLoaded(true);
            }
          
          })
        }
      })
      .catch((error) => {
        console.log("Error getting documents: ", error);
      });
    };

  
    fetchList();  
  }, []);

  
  return (

    <View style={styles.app}>
       {
         isLoaded ? 
         <View style={styles.fullWidthWindow}>
           <View style={styles.floatingActionView}>
             <Image source={{ uri: item.imagePath ? item.imagePath : Image.resolveAssetSource(defaultImage).uri }}
               style={styles.challengePhoto}></Image>
             <SetFeatured firebaseCollectionName="challenges" item={item} />
             <Tags difficulty={item.difficulty} topics={item.topics}></Tags>
             <Text style={styles.contentTitle}>{item.title}</Text> 
             <Text style={styles.contentDesc}>{item.description}{"\n"}</Text>
             <Text>Challenge starts on: {theStartDate}</Text>
             <Text>Challenge ends on: {theEndDate}</Text>
             <Text>This challenge is {item.numDays} day(s) long {"\n"}</Text>
             <ChallengeDayList 
               contentComponent="ChallengeDay"
               navigation={navigation}
               contentType="challengeDays"
               data={data.sort((docA, docB) => docB.dateAdded - docA.dateAdded)} />
           </View>
         </View>
         : <LoadingSpinner />
     }
     </View>
 
  );
};

export default Challenge;
