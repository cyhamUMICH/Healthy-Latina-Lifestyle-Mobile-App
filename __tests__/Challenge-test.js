// Code from: https://github.com/byCedric/expo-guide-ci/blob/master/__tests__/App-test.js
//      and https://jestjs.io/docs/tutorial-react-native
import React from 'react';
import Challenge from '../src/screens/Challenge';
import renderer from 'react-test-renderer';

const mockedParams = {
  route: { 
    params: {
      "contentID": "jpu5HUQLthxhkjo7HE1i",
      "dateAdded": {
        "nanoseconds": 0,
        "seconds": 1620403200,
      },
      "startDate": {
        "nanoseconds": 0,
        "seconds": 1622520000,
      },
      "endDate": {
        "nanoseconds": 0,
        "seconds": 1625111999,
      },
      "description": "Challenges for the month of June.",
      "difficulty": "A",
      "featured": true,
      "imagePath": "",
      "language": "EN",
      "title": "June Challenge",
      "topics": [
        "Anxiety"
      ]
    }
  }
};

// https://stackoverflow.com/questions/66563806/typeerror-cannot-read-property-params-of-undefined-jest-testing-react-nativ
test('Renders Challenge With the Loading Screen', async () => {
  // const tree = renderer.create(<Challenge {...mockedParams} />).toJSON();
  // expect(tree).toMatchSnapshot();
  console.log("Renders Challenge With the Loading Screen: This test is disabled.");
});

test('Renders Challenge Without the Loading Screen', async () => {
  // const tree = renderer.create(<Challenge {...mockedParams} skipLoadingScreen />).toJSON();
  // expect(tree).toMatchSnapshot();
  console.log("Renders Challenge Without the Loading Screen: This test is disabled.");
});