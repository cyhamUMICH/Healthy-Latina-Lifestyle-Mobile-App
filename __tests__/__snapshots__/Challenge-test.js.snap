// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Challenge With the Loading Screen 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingHorizontal": "2%",
      "paddingTop": "2%",
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://healthylatinalifestylecom.files.wordpress.com/2017/10/cropped-cropped-peackcock-color-healthy-latino-20171.jpg",
      }
    }
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 200,
        "height": 300,
        "justifyContent": "center",
        "left": 50,
        "width": 300,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "marginTop": "2%",
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "#89cff0",
          "borderRadius": 20,
          "marginBottom": "2%",
          "marginRight": "2%",
          "paddingHorizontal": "5%",
          "paddingVertical": "2%",
        }
      }
    >
      Advanced
    </Text>
    <Text
      style={
        Object {
          "backgroundColor": "#d7efc0",
          "borderRadius": 20,
          "marginBottom": "2%",
          "marginRight": "2%",
          "paddingHorizontal": "5%",
          "paddingVertical": "2%",
        }
      }
    >
      Anxiety
    </Text>
  </View>
  <Text
    style={
      Object {
        "fontSize": 28,
        "textAlign": "center",
      }
    }
  >
    June Challenge
  </Text>
  <RCTScrollView
    style={
      Object {
        "marginBottom": "5%",
        "marginTop": "2%",
      }
    }
  >
    <View>
      <Text>
        Challenges for the month of June.
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Renders Challenge Without the Loading Screen 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingHorizontal": "2%",
      "paddingTop": "2%",
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://healthylatinalifestylecom.files.wordpress.com/2017/10/cropped-cropped-peackcock-color-healthy-latino-20171.jpg",
      }
    }
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 200,
        "height": 300,
        "justifyContent": "center",
        "left": 50,
        "width": 300,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "marginTop": "2%",
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "#89cff0",
          "borderRadius": 20,
          "marginBottom": "2%",
          "marginRight": "2%",
          "paddingHorizontal": "5%",
          "paddingVertical": "2%",
        }
      }
    >
      Advanced
    </Text>
    <Text
      style={
        Object {
          "backgroundColor": "#d7efc0",
          "borderRadius": 20,
          "marginBottom": "2%",
          "marginRight": "2%",
          "paddingHorizontal": "5%",
          "paddingVertical": "2%",
        }
      }
    >
      Anxiety
    </Text>
  </View>
  <Text
    style={
      Object {
        "fontSize": 28,
        "textAlign": "center",
      }
    }
  >
    June Challenge
  </Text>
  <RCTScrollView
    style={
      Object {
        "marginBottom": "5%",
        "marginTop": "2%",
      }
    }
  >
    <View>
      <Text>
        Challenges for the month of June.
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
